physx/src/query.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+crate::DeriveClassForNewType!(PxQueryFilterData: PxQueryFilterData);
+
+impl Default for PxQueryFilterData {
+    fn default() -> Self {
+        unsafe { physx_sys::PxQueryFilterData_new().into() }
+    }
+}
+
+impl From<physx_sys::PxQueryFilterData> for PxQueryFilterData {
+    fn from(data: physx_sys::PxQueryFilterData) -> Self {
+        PxQueryFilterData { obj: data }
+    }
+}
+
+impl Into<physx_sys::PxQueryFilterData> for PxQueryFilterData {
+    fn into(self) -> physx_sys::PxQueryFilterData {
+        self.obj
+    }
+}
+
+#[repr(transparent)]
+pub struct PxRaycastCallback {
+    pub obj: Owner<physx_sys::PxRaycastCallback>,
+}
+
+#[repr(transparent)]
+pub struct PxRaycastBuffer(PxRaycastCallback);
+
+unsafe impl Class<PxRaycastCallback> for PxRaycastBuffer {
+    fn as_ptr(&self) -> *const PxRaycastCallback {
+        self.0.obj.as_ptr() as *const _ as *const _
+    }
+    fn as_mut_ptr(&mut self) -> *mut PxRaycastCallback {
+        self.0.obj.as_mut_ptr() as *mut _ as *mut _
+    }
+}
+
+unsafe impl Class<physx_sys::PxRaycastCallback> for PxRaycastBuffer {
+    fn as_ptr(&self) -> *const physx_sys::PxRaycastCallback {
+        self.0.obj.as_ptr() as *const _ as *const _
+    }
+    fn as_mut_ptr(&mut self) -> *mut physx_sys::PxRaycastCallback {
+        self.0.obj.as_mut_ptr() as *mut _ as *mut _
+    }
+}
+
+unsafe impl RaycastCallback for PxRaycastBuffer {
+    unsafe fn new() -> Self {
+        Self(PxRaycastCallback {
+            obj: Owner::from_raw(create_raycast_buffer()).unwrap(),
+        })
+    }
+
+    unsafe fn get_raycast_hit(&mut self) -> Option<physx_sys::PxRaycastHit> {
+        Some(self.0.obj.block)
+    }
+
+    unsafe extern "C" fn process_touches(
+        _buffer: *const physx_sys::PxRaycastHit,
+        _num_hits: u32,
+        _user_data: *const c_void,
+    ) -> bool {
+        // Unused by this impl
+        false
+    }
+
+    unsafe extern "C" fn finalize_query(_user_data: *const c_void) {
+        // Unused by this impl
+    }
+
+    unsafe fn into_px(&mut self) -> *mut physx_sys::PxRaycastCallback {
+        self.0.obj.as_mut_ptr()
+    }
+}
+
+/// A trait for creating custom raycast callbacks for PhysX.
+pub unsafe trait RaycastCallback: Sized {
+    unsafe fn new() -> Self;
+    unsafe fn get_raycast_hit(&mut self) -> Option<physx_sys::PxRaycastHit>;
+
+    /// # Safety
+    /// Must not panic.
+    unsafe extern "C" fn process_touches(
+        buffer: *const physx_sys::PxRaycastHit,
+        num_hits: u32,
+        user_data: *const c_void,
+    ) -> bool;
+
+    /// # Safety
+    /// Must not panic.
+    unsafe extern "C" fn finalize_query(user_data: *const c_void);
+
+    /// # Safety
+    /// Do not override this method.
+    unsafe fn into_px(&mut self) -> *mut physx_sys::PxRaycastCallback {
+        create_raycast_callback(
+            Self::process_touches,
+            Self::finalize_query,
+            Box::into_raw(Box::new(self)) as *mut c_void,
+        )
+    }
```

Comment:
Looking at this again, I don't think that `into_px` function is correct, `Scene::raycast` is creating a `physx_sys::PxRaycastCallback` every time it's called.  The `user_data` also needs to be cleaned up.
The starting point is to make a new-type wrapper with a generic, but it's gonna get a little tricky.
```Rust
pub struct PxRaycastCallback<T: RaycastCallback> {
    obj: physx_sys::PxRaycastCallback,
    phantom_user_data: PhantomData<T>,
}

impl<T: RaycastCallback> PxRaycastCallback<T> {
    fn new(user_data: T) -> Option<Owner<Self>> {
        create_raycast_callback(
            Self::process_touches,
            Self::finalize_query,
            Bow::into_raw(Box::new(user_data)), // we need to make sure this is dropped
        )
    }

    unsafe extern "C" fn process_touches(
        buffer: *const physx_sys::PxOverlapHit,
        num_hits: u32,
        user_data: *const c_void,
    ) -> bool {
        (&*(user_data as *const T)).process_callbacks(slice::from_raw(buffer, num_hits as usize))
    }
    // finalize_query wrapper as well
}
```
The drop impl will need a `user_data` getter and a dtor for the rest of it in C++, and that will need to be integrated in a way that doesn't allow calling something like `PxRaycastBuffer::get_user_data(...)` because it lacks the trampoline fields.  `into_px` should consume `self` and produce `PxRaycastCallback<Self>`.  Some methods/traits may need to be shifted around so that there's a sensible and safe trait bound exposed in `Scene::raycast`.

User data makes everything messy 😢

Reply:
`physx_sys::PxRaycastHit` (and co.) should also get safe wrappers, that could be done in a follow up though.